web/source/text/defaultOutput.ts

```diff
@@ -11,68 +11,118 @@ namespace com.keyman.text {
     private constructor() {
     }
 
+    static codeForEvent(Lkc: KeyEvent) {
+      let keyName = Lkc.kName;
+      let n = Lkc.Lcode;
+
+      var code = Codes.keyCodes[keyName];
+      if(!code) {
+        code = n;
+      }
+
+      return code;
+    }
+
+    /**
+     * Serves as a default keycode lookup table.  This may be referenced safely by mnemonic handling without fear of side-effects.
+     * Also used by Processor.defaultRuleBehavior to generate output after filtering for special cases.
+     */
+    public static forAny(Lkc: KeyEvent, shiftState: number) {
+      var char = '';
+
+      let code = DefaultOutput.codeForEvent(Lkc);
+
+      // A pretty simple table of lookups, corresponding VERY closely to the original defaultKeyOutput.
+      if(char = DefaultOutput.forSpecialEmulation(Lkc)) {
+        return char;
+      } else if(char = DefaultOutput.forNumpadKeys(Lkc)) {
+        return char;
+      } else if(char = DefaultOutput.forUnicodeKeynames(Lkc)) {
+        return char;
+      } else if(char = DefaultOutput.forBaseKeys(Lkc, shiftState)) {
+        return char;
+      } else {
+        // // For headless and embeddded, we may well allow '\t'.  It's DOM mode that has other uses.
+        // // Not originally defined for text output within defaultKeyOutput.
+        // // We can't enable it yet, as it'll cause hardware keystrokes in the DOM to output '\t' rather
+        // // than rely on the browser-default handling.
+        // switch(code) {
+        //   case Codes.keyCodes['K_TAB']:
+        //   case Codes.keyCodes['K_TABBACK']:
+        //   case Codes.keyCodes['K_TABFWD']:
+        //     return '\t';
+        //   default:
+        //     return '';
+        // }
+      }
+    }
+
+    public static isDOMCommand(Lkc: KeyEvent): boolean {
+      let code = DefaultOutput.codeForEvent(Lkc);
+
+      switch(code) {
+        case Codes.keyCodes['K_TAB']:
+        case Codes.keyCodes['K_TABBACK']:
+        case Codes.keyCodes['K_TABFWD']:
+          return true;
+        default:
+          return false;
+      }
+    }
+
+    /**
+     * forCommands - returns a boolean indicating if a non-text event should be triggered by the keystroke.
+     */
+    public static isCommand(Lkc: KeyEvent): boolean {
```

Comment:
I'm establishing the pattern now so that when the DOM ties are fully cut, the review process then will be simpler (just code relocation) and far more straightforward than requiring logic edits and analysis as well.  Also note `applyCommand` vs `applyDOMCommand`.

If naught else, it helps me feel that this aspect is fully prepared, rather than leave another detail lingering to be redone at a later point.  There are a lot of moving pieces yet to be handled.